tf2_ros/src/static_transform_broadcaster_program.cpp

```diff
@@ -35,68 +35,172 @@
+  }
+  /* otherwise, look for roll, pitch, yaw */
+  double roll = 0.0;
+  double pitch = 0.0;
+  double yaw = 0.0;
+  iter = args.find("--roll");
+  if (iter != args.end()) {
+    roll = std::stod(iter->second);
+    found_rpy = true;
+  }
+  iter = args.find("--pitch");
+  if (iter != args.end()) {
+    pitch = std::stod(iter->second);
+    found_rpy = true;
+  }
+  iter = args.find("--yaw");
+  if (iter != args.end()) {
+    yaw = std::stod(iter->second);
+    found_rpy = true;
+  }
+  if (found_quaternion && found_rpy) {
+    RCUTILS_LOG_ERROR("cannot mix euler and quaternion arguments");
+    _print_usage();
+    exit(1);
+  } else if (!found_quaternion) {
+    quat.setRPY(roll, pitch, yaw);
+  }
+  return quat;
+}
+
+tf2::Vector3 _get_translation(const std::unordered_map<std::string, std::string> & args)
+{
+  tf2::Vector3 trans;
+  auto iter = args.find("--x");
+  if (iter != args.end()) {
+    trans.setX(std::stod(iter->second));
+  }
+  iter = args.find("--y");
+  if (iter != args.end()) {
+    trans.setY(std::stod(iter->second));
+  }
+  iter = args.find("--z");
+  if (iter != args.end()) {
+    trans.setZ(std::stod(iter->second));
+  }
+  return trans;
+}
+
 int main(int argc, char ** argv)
 {
   // Initialize ROS
   std::vector<std::string> args = rclcpp::init_and_remove_ros_arguments(argc, argv);
+  std::unordered_map<std::string, std::string> arg_map = _make_arg_map(std::move(args));
   rclcpp::NodeOptions options;
   std::shared_ptr<tf2_ros::StaticTransformBroadcasterNode> node;
-
-  if (args.size() != 9 && args.size() != 10) {
-    printf("A command line utility for manually sending a transform.\n");
-    printf("Usage: static_transform_publisher x y z qx qy qz qw frame_id child_frame_id \n");
-    printf("OR \n");
-    printf("Usage: static_transform_publisher x y z yaw pitch roll frame_id child_frame_id \n");
-    RCUTILS_LOG_ERROR(
-      "static_transform_publisher exited due to not having the right number of arguments");
-    return 2;
+  tf2::Quaternion rotation;
+  tf2::Vector3 translation;
+  try {
+    rotation = _get_rotation(arg_map);
+    translation = _get_translation(arg_map);
+  } catch (std::invalid_argument & e) {
+    RCUTILS_LOG_ERROR("error parsing command line arguments");
+    _print_usage();
+    return 1;
   }
-  double x = std::stod(args[1]);
-  double y = std::stod(args[2]);
-  double z = std::stod(args[3]);
-  double rx, ry, rz, rw;
   std::string frame_id, child_id;
-
-  if (args.size() == 9) {
-    // grab parameters from yaw, pitch, roll
-    tf2::Quaternion quat;
-    quat.setRPY(std::stod(args[6]), std::stod(args[5]), std::stod(args[4]));
-    rx = quat.x();
-    ry = quat.y();
-    rz = quat.z();
-    rw = quat.w();
-    frame_id = args[7];
-    child_id = args[8];
-  } else {
-    // quaternion supplied directly
-    rx = std::stod(args[4]);
-    ry = std::stod(args[5]);
-    rz = std::stod(args[6]);
-    rw = std::stod(args[7]);
-    frame_id = args[8];
-    child_id = args[9];
```

Comment:
I think we still need to keep this old command-line parsing around for backwards-compatibility reasons.  My suggestion is that we first look for the new arguments; as we find them in the map, we remove them from the map.  Once we are done removing the new options, we look to see if there are 9 or 10 arguments left.  If there are, we print a warning that this usage is deprecated, but continue on anyway.  If both the new arguments and the old arguments are provided, we error and exit.  Does that seem like a reasonable plan?

Reply:
Yeah, that's absolutely a reasonable plan. Will do.